src/Features/Core/Portable/Completion/CompletionServiceWithProviders.cs

```diff
@@ -283,6 +291,33 @@ private CompletionProvider GetProviderByName(string providerName)
                 (expandItemsAvailableFromTriggeredProviders || expandItemsAvailableFromAugmentingProviders));
         }
 
+        private bool ValidatePossibleTriggerCharacterSet(IEnumerable<CompletionProvider> triggeredProviders, Document document, SourceText text, int caretPosition)
+        {
+            var syntaxFactsService = document.Project.LanguageServices.GetService<ISyntaxFactsService>();
+            if (caretPosition > 0 && syntaxFactsService != null)
+            {
+                // The trigger character has already been inserted before the current caret position.
+                var character = text[caretPosition - 1];
+
+                // Identifier characters are not part of the possible trigger character set, so don't validate them.
+                var isIdentifierCharacter = syntaxFactsService.IsIdentifierStartCharacter(character) || syntaxFactsService.IsIdentifierEscapeCharacter(character);
+                if (isIdentifierCharacter)
+                {
+                    return true;
+                }
+
+                // Only verify against built in providers.  3rd party ones do not necessarily implement the possible trigger characters API.
+                triggeredProviders = triggeredProviders.Where(p => GetBuiltInProviders().Contains(p));
+                foreach (var provider in triggeredProviders)
+                {
+                    Debug.Assert(provider.PossibleTriggerCharacters.Contains(character),
+                        $"the character {character} is not a valid trigger character for {provider.Name}");
+                }
+            }
+
+            return true;
+        }
```

Comment:
i'm really glad you added this.  i very much want us tohave mechanisms to catch misbehaving providers.